### ISSUE TITLE
BaseTimeSeries problem with getSubList

### ISSUE THREAD

**Issue by MetalMiner** [ta4j/ta4j#324]
Hey guys,

I have a BaseTimeSeries where I set the MaximumBarCount for (in my example to ). Every minute there is a new Bar added, When I call the getSubList method after a Bar was added with the follwing statement:

`timeSeries.getSubSeries(timeSeries.getEndIndex() - (timespan + 1), timeSeries.getEndIndex())`

an Exception is thrown:

`java.lang.IndexOutOfBoundsException: toIndex = 40321
	at java.util.ArrayList.subListRangeCheck(ArrayList.java:1012)
	at java.util.ArrayList.subList(ArrayList.java:1004)
	at org.ta4j.core.BaseTimeSeries.cut(BaseTimeSeries.java:423)
	at org.ta4j.core.BaseTimeSeries.getSubSeries(BaseTimeSeries.java:199)
	at de.coin.research.service.impl.Ta4JTimeSeriesServiceImpl.getTimeSeries(Ta4JTimeSeriesServiceImpl.java:84)
	at de.coin.research.service.impl.Ta4JTimeSeriesServiceImpl$$FastClassBySpringCGLIB$$70f1fbc.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:746)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:88)
	at de.coin.research.aop.logging.LoggingAspect.logAround(LoggingAspect.java:85)
	at sun.reflect.GeneratedMethodAccessor344.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:644)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:633)
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.aspectj.AspectJAfterThrowingAdvice.invoke(AspectJAfterThrowingAdvice.java:62)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at de.coin.research.service.impl.Ta4JTimeSeriesServiceImpl$$EnhancerBySpringCGLIB$$c7a0d204.getTimeSeries(<generated>)
	at de.coin.research.service.mapper.Ta4JMapper.getTimeSeries(Ta4JMapper.java:239)
	at de.coin.research.service.mapper.Ta4JMapper.mapToIndicator(Ta4JMapper.java:85)
	at de.coin.research.service.mapper.Ta4JMapper.mapToIndicator(Ta4JMapper.java:81)
	at de.coin.research.service.mapper.Ta4JMapper.getIndicatorFromConfig(Ta4JMapper.java:546)
	at de.coin.research.service.mapper.Ta4JMapper.mapToRule(Ta4JMapper.java:261)
	at de.coin.research.service.mapper.Ta4JMapper.mapToRule(Ta4JMapper.java:246)
	at de.coin.research.service.mapper.Ta4JMapper.mapToStrategy(Ta4JMapper.java:517)
	at de.coin.research.service.mapper.Ta4JMapper$$FastClassBySpringCGLIB$$af6a4255.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:746)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:294)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at de.coin.research.service.mapper.Ta4JMapper$$EnhancerBySpringCGLIB$$a72f5c54.mapToStrategy(<generated>)
	at de.coin.research.business.StrategyRunner.runStrategies(StrategyRunner.java:59)
	at de.coin.research.business.StrategyRunner$$FastClassBySpringCGLIB$$bebd3d4e.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:746)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:294)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at de.coin.research.business.StrategyRunner$$EnhancerBySpringCGLIB$$ee1c70df.runStrategies(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)`

The problem is, that when the BaseTimeSeries exceed the MaximumBarCount it removes the oldest Bar (which is correct) but raises the endIndex by 1 (which is wrong I think). This leads to the index out of Bounds error when you call the getSubList mehtod via endIndex.

**Comment by team172011** [comment]
Hi @MetalMiner, I think you are right and this is a bug. Please provide a working example code next time ;-)

```java
    @Test
    public void SubSeriesOfMaxBarCountSeriesTest() {
    	final TimeSeries series = new BaseTimeSeries.SeriesBuilder()
    							.withNumTypeOf(numFunction)
    							.withName("Series with maxBar count")
    							.withMaxBarCount(20)
    							.build();
    	final int timespan = 5;
    	
    	IntStream.range(0, 100).forEach(i -> {
    		series.addBar(ZonedDateTime.now(ZoneId.systemDefault()).plusMinutes(i),5,7,1,5,5);
    		TimeSeries subseries = series.getSubSeries(series.getEndIndex() - (timespan + 1), series.getEndIndex());
    		System.out.println(String.format("Subseries #%s of size %s created.",i,subseries.getBarCount()));
    		});
    }
```

```java
java.lang.IndexOutOfBoundsException: toIndex = 21
	at java.util.ArrayList.subListRangeCheck(ArrayList.java:1012)
	at java.util.ArrayList.subList(ArrayList.java:1004)
	at org.ta4j.core.BaseTimeSeries.cut(BaseTimeSeries.java:423)
	at org.ta4j.core.BaseTimeSeries.getSubSeries(BaseTimeSeries.java:199)
	at org.ta4j.core.TimeSeriesTest.lambda$0(TimeSeriesTest.java:312)
	at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110)
	at java.util.stream.IntPipeline$Head.forEach(IntPipeline.java:557)
	at org.ta4j.core.TimeSeriesTest.SubSeriesOfMaxBarCountSeriesTest(TimeSeriesTest.java:310)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)

```

**Comment by team172011** [comment]
This will be fixed with PR #325. 
I am thinking about to remove the `setMaximumBarCount ` method and only to allow to determine this parameter in constructor/builder

**Comment by MetalMiner** [comment]
Thanks for the quick response. I apprechiate your solution!

**Comment by milczarekIT** [comment]
Fixed in #474

**Comment by nimo23** [comment]
@team172011 if this is fixed, then plz close this issue.
